# Task Rabbit Data Analysis
by Zack Seyun Kim

In [2]:
%%HTML
<script>
  function code_toggle() {
    if (code_shown){
      $('div.input').hide('500');
      $('#toggleButton').val('Show Code')
    } else {
      $('div.input').show('500');
      $('#toggleButton').val('Hide Code')
    }
    code_shown = !code_shown
  }

  $( document ).ready(function(){
    code_shown=false;
    $('div.input').hide()
  });
</script>
<form action="javascript:code_toggle()"><input type="submit" id="toggleButton" value="Show Code"></form>

In [3]:
import pandas as pd 
import numpy as np

In [4]:
data = pd.read_csv('sample.csv')

### 'A "recommendation" is the group of Taskers from which the Client can choose one to book.'


In [4]:
#provided dataset:
data.head()

recommendation_id           created_at   tasker_id  \
0  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1009185352   
1  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1006892359   
2  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1012023956   
3  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1009733517   
4  0-0-70cf97d7-37af-4834-901c-ce3ad4893b8c  2017-09-01 00:32:25  1013579273   

   position  hourly_rate  num_completed_tasks  hired            category  
0         1           38                  151      0  Furniture Assembly  
1         2           40                  193      0  Furniture Assembly  
2         3           28                    0      0  Furniture Assembly  
3         4           43                  303      0  Furniture Assembly  
4         5           29                   39      0  Furniture Assembly

In [7]:
data.describe()

tasker_id      position   hourly_rate  num_completed_tasks  \
count  3.000000e+04  30000.000000  30000.000000         30000.000000   
mean   1.011072e+09      7.873600     57.482533           221.235700   
std    2.501822e+06      4.319442     36.222860           279.801812   
min    1.006647e+09      1.000000     18.000000             0.000000   
25%    1.008955e+09      4.000000     38.000000            23.000000   
50%    1.010579e+09      8.000000     45.000000           114.000000   
75%    1.013444e+09     12.000000     60.000000           300.250000   
max    1.015020e+09     15.000000    290.000000          1406.000000   

              hired  
count  30000.000000  
mean       0.056833  
std        0.231528  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [6]:
recSets = data.groupby('recommendation_id')

1.) **How many recommendation sets are in this data sample?**<br>
2,100


In [7]:
len(recSets)

2100

In [8]:
#either way to find unique recSets (using groupby or unique count)
data.recommendation_id.nunique()

2100

2.) **Each recommendation set shows from 1 to 15 Taskers, what is:**
     - average number of Taskers shown for each recommendation set: 14.3
     - median  number of Taskers shown for each recommendation set: 15
   

In [23]:
#because we grouped by recommendation_id set earlier, we just need to iterate through the pandas group object
import statistics
setTaskers = []
for rec_id,Set in recSets:
    setTaskers.append(len(Set))
print(statistics.mean(setTaskers))
print(statistics.median(setTaskers))
    

14.285714285714286
15.0


3.**How many total unique Taskers are there in this data sample?**<br>
830

In [27]:
data.tasker_id.nunique()

830

4.) **Which Tasker has been shown the most?**<br>Tasker:1014508755 has been shown in the most recommendation sets. with 608 impressions<br>
   **Which Tasker has been shown the least?**<br> Tasker: 1013573988 et al. the least shown taskers with only one impression in the recommendation sets(see below)

In [57]:
#**Which Tasker has been shown the most?**
data.tasker_id.value_counts().head()

1014508755    608
1012043028    438
1014675294    387
1014629676    311
1015015238    290
Name: tasker_id, dtype: int64

In [19]:
#Which Tasker has been shown the least?
shown = pd.DataFrame(data.tasker_id.value_counts())
shown.loc[shown['tasker_id']==1]

tasker_id
1013573988          1
1009994950          1
1012386513          1
1007383273          1
1014926743          1
1009112003          1
1010009736          1
1014310300          1
1012678504          1
1012364558          1
1012166729          1
1010021990          1
1013573125          1
1011901532          1
1011949117          1
1010681878          1
1011968845          1
1013830691          1
1012348656          1
1009871933          1
1012289475          1
1006853970          1
1013934937          1
1006690425          1
1008469117          1
1009641175          1
1014547884          1
1010779242          1
1008604368          1
1009603880          1
...               ...
1009772528          1
1014478773          1
1009754999          1
1008919567          1
1008368716          1
1013854788          1
1013656032          1
1013731883          1
1013362004          1
1010042971          1
1007246122          1
1011952623          1
1009547227          1
1008033678          1
1012805440          1
1007472083          1
1009618500          1
1011985968          1
1009702351          1
1007638825          1
1009712638          1
1007480912          1
1008870833          1
1010640007          1
1014593279          1
1014086818          1
1008474216          1
1012151299          1
1014439502          1
1011972750          1

[68 rows x 1 columns]

5.) **Which Tasker has been hired the most?**<br>
user: 1012043028 with 59 hires (found by querying those who've been hired, and using value_counts function to aggregate unique_taskers with most hires <br>
**Which Tasker has been hired the least?**<br>1006646767 et al. group of taskers(found by grouping table by tasker_id, and using sum as the aggregate function on the hired column. the taskers with 0 hires have been hired the least.(see cell below)

In [39]:
#**Which Tasker has been hired the most?**
hired = data.loc[data['hired']==1].tasker_id.value_counts()
hired[:5]

1012043028    59
1013131759    39
1013359522    37
1013165984    36
1013794735    35
Name: tasker_id, dtype: int64

In [16]:
#**Which Tasker has been hired the least?**
least = pd.DataFrame(data.groupby('tasker_id').hired.sum())
least.loc[least['hired']==0]

hired
tasker_id        
1006646767      0
1006655883      0
1006690425      0
1006702141      0
1006720473      0
1006751673      0
1006771484      0
1006797028      0
1006808958      0
1006853970      0
1006899551      0
1006912729      0
1006997279      0
1007007551      0
1007030973      0
1007062502      0
1007079569      0
1007084219      0
1007099706      0
1007184581      0
1007227664      0
1007246122      0
1007281889      0
1007295623      0
1007298334      0
1007306634      0
1007330127      0
1007341779      0
1007359134      0
1007383273      0
...           ...
1014532398      0
1014539109      0
1014547884      0
1014551627      0
1014571994      0
1014578899      0
1014586046      0
1014593279      0
1014632169      0
1014640927      0
1014657461      0
1014659781      0
1014667352      0
1014697426      0
1014710322      0
1014753188      0
1014753677      0
1014788707      0
1014790265      0
1014820493      0
1014860648      0
1014866421      0
1014881086      0
1014901856      0
1014917205      0
1014926743      0
1014928212      0
1014983247      0
1015000099      0
1015000442      0

[518 rows x 1 columns]

6.) **If we define the "Tasker conversion rate" as the number of times a
   Tasker has been hired, out of the number of times the Tasker has
   been shown, how many Taskers have a conversion rate of 100%**<br> 6 taskers had 100% conversion rate

In [69]:
shown = data.tasker_id.value_counts()
hired = data.groupby('tasker_id').hired.sum()

In [73]:
#**the 'tasker_id' column in the following dataframe represents the number of times the tasker_id has been shown**
convRate = pd.concat([hired,shown],axis=1)
convRate['rate'] = convRate['hired']/convRate['tasker_id']

In [76]:
hundred = convRate.loc[convRate['rate']==1]
hundred

hired  tasker_id  rate
1007480912      1          1   1.0
1008094420      2          2   1.0
1008861741      9          9   1.0
1011985968      1          1   1.0
1012369686      2          2   1.0
1014478773      1          1   1.0

In [77]:
len(hundred)

6

7.) **Would it be possible for all Taskers to have a conversion rate of 100%
   Please explain your reasoning.**<br>No that is not be possible. Each recommendation set include a group of taskers in which there is one person that is hirable. In order to offer the user-employer a selection of possible taskers, there is a larger supply of taskers available to the market than is being hired at any given time. Therefore 14/15 taskers in a recommendation set will inevitably have lower conversion rates in every group the taskers are included in. 

8.) **For each category, what is the average position of the Tasker who is hired?**<br> Answered the average and mode (most frequently selected)of the numerical position per category as the data was useful.

In [8]:
data.groupby('category').position.mean()

category
Furniture Assembly    7.8801
Mounting              7.8706
Moving Help           7.8701
Name: position, dtype: float64

#### Appears that the most hired are in within the early selections

In [7]:
data.groupby('category').position.agg(pd.Series.mode)

category
Furniture Assembly         1
Mounting                   1
Moving Help           [1, 2]
Name: position, dtype: object

9.) **For each category, what is the average hourly rate and average number of completed tasks for the Taskers who are hired?**

In [82]:
#avereage hourly rate
data.loc[data['hired']==1].groupby('category').hourly_rate.mean()

category
Furniture Assembly    38.701049
Mounting              50.154804
Moving Help           63.012259
Name: hourly_rate, dtype: float64

In [83]:
#average number of completed tasks
data.loc[data['hired']==1].groupby('category').num_completed_tasks.mean()

category
Furniture Assembly    249.020979
Mounting              284.096085
Moving Help           273.882662
Name: num_completed_tasks, dtype: float64

10.) **Based on the previous, how would you approach the question of:
    How can we use market data to suggest hourly rates to Taskers that would
    maximize their opportunity to be hired?
    Please describe in detail, with code and formulas that support your model.**

* The easiest method of practical value, is to use first part of question 9: to group by category then recommendation-set. Then average the hourly_rates of those that managed to be hired, suggesting the range: mean(plus minus one standard deviation) is the range of hourly rate that has worked best with this category in the past. 
* Possibly, another method would be to attempt a statistical model (decision tree-based or something more simple such as linear regression). By transforming categorical data into dummy variabels, we can create a linear regression model with input variables of position,num_completed tasks, and dummy_categories that can regressively predict what the hourly rate should be based on past data of hired taskers. 
* The limitation with the 2nd method is that it doesnt take advantage of the data of those that weren't hired. Also the features available may be insufficient in determining optimal price. 


### Method 1

In [10]:
data.loc[data['hired']==1].groupby('category').hourly_rate.std()

category
Furniture Assembly    13.514984
Mounting              10.043536
Moving Help           29.959649
Name: hourly_rate, dtype: float64

Furniture Assembly----38.701049 plus or minus 13.514984<br>
Mounting---------------50.154804 plus or minus 10.043536<br>
Moving Help-----------63.012259 plus or minus 29.959649<br>

###  Method 2

In [121]:
furniture = data.loc[(data['hired']==1)]

In [91]:
#create dummy variables for the categories
dummy = pd.get_dummies(furniture, columns=['category'],drop_first=True)

**R2 Value Linear Regression:**

In [95]:
X = pd.concat([furniture.iloc[:,3],furniture.iloc[:,5],dummy.iloc[:,-2:]], axis=1)
y = furniture['hourly_rate'].reset_index(drop= True)

#K Fold Cross validation implementation
from sklearn.linear_model import LinearRegression
from sklearn import metrics
from sklearn.model_selection import cross_val_predict
lr = LinearRegression()
predicted = cross_val_predict(lr, X, y, cv=10)
print(metrics.r2_score(y, predicted))

0.3942976733109893


In [96]:
print(predicted.min(), predicted.mean(),predicted.max())

28.380302441641504 50.55620237914096 96.16436950699025


#### Random Forest Model

Random Forest R2

In [116]:
from sklearn.ensemble import RandomForestRegressor
forest = RandomForestRegressor(n_estimators=10, 
                                criterion='mse', 
                                random_state=1, 
                                n_jobs=1)
forest.fit(X, y)
pred = cross_val_predict(forest, X, y, cv=10)
forest.score(X,y)

0.927404945849187

Random Forest was able to give a much stronger prediction. This hints our features are sufficient in predicting a good hourly wage despite limited input features. 

(first line is RF Model predicted values, second vertical array is actual wage data)

In [120]:
print(pred[:10])
print(y[:10])

[44.5  45.   33.86 99.   32.3  31.8  42.5  42.5  42.5  71.7 ]
0    50
1    50
2    32
3    95
4    35
5    35
6    42
7    42
8    42
9    34
Name: hourly_rate, dtype: int64


Although inear regression weakly predicted the response variable, random forest was able to perform much better in suggesting a valid a wage that will most likely help a tasker get hired.

#### To be noted
Of course,  these models are extremely limited and may not reflect recent studies done about profile selection(hirability in our case). Similar to studies done for evaluating Tinder Profiles: Profile Picture, Description, and Rating may affect selection decision more than price which may be more elastic pending on employer. Therefore coding the extra data aforementioned(making it usable input for a model)will more than likely lead to more conclusive and valuable results. 